envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/ServiceTagFilter.kt

```diff
@@ -0,0 +1,129 @@
+package pl.allegro.tech.servicemesh.envoycontrol.snapshot
+
+class ServiceTagFilter(properties: ServiceTagsProperties = ServiceTagsProperties()) {
+
+    private val tagsBlacklist: List<Regex> = properties.routingExcludedTags.map { Regex(it) }
+    private val twoTagsCombinationsByService: Map<String, List<Pair<Regex, Regex>>>
+    private val threeTagsCombinationsByService: Map<String, List<Triple<Regex, Regex, Regex>>>
+
+    init {
+        properties.allowedTagsCombinations.forEach {
+            if (it.tags.size < 2 || it.tags.size > 3) {
+                throw IllegalArgumentException(
+                    "A tags combination must contain 2 or 3 tags. Combination with ${it.tags.size} tags found")
+            }
+        }
+        val combinationsByService = properties.allowedTagsCombinations
+            .groupBy({ it.serviceName }) { it.tags }
+
+        threeTagsCombinationsByService = combinationsByService
+            .mapValues { it.value.filter { it.size == 3 } }
+            .mapValues { it.value
+                .map { it.sorted().map { Regex(it) } }
+                .map { Triple(it[0], it[1], it[2]) }
+                .distinct()
+            }
+            .filterValues { it.isNotEmpty() }
+
+        twoTagsCombinationsByService = combinationsByService
+            .mapValues { it.value.filter { it.size == 2 } }
+            .mapValues { it.value
+                .map { it.sorted().map { Regex(it) } }
+                .map { it[0] to it[1] }
+                .distinct()
+            }
+            .mapValues { (serviceName, combinations) ->
+                combinations + threeTagsCombinationsByService[serviceName].orEmpty()
+                    .flatMap { it.getAllPairs() }
+            }
+            .filterValues { it.isNotEmpty() }
+    }
+
+    /**
+     * Transforms raw instance's tags to tags that should actually be add to instance metadata.
+     *  - discards tags that are not suitable for routing
+     *  - generates tags combinations that may be used for routing
+     *
+     * @return sequence of tags that should be add to instance's metadata or null if tag entry should not be add to
+     * instance's metadata.
+     */
+    fun getAllTagsForRouting(serviceName: String, instanceTags: Set<String>): Sequence<String>? {
+        val tags = filterTagsForRouting(instanceTags)
+        if (tags.isEmpty()) {
+            return null
+        }
+
+        val addPairs = isAllowedToMatchOnTwoTags(serviceName)
+        val addTriples = isAllowedToMatchOnThreeTags(serviceName)
+        val generatePairs = addPairs || addTriples
+
+        val tagsPairs = if (generatePairs) {
+            tags
+                .flatMap { tag1 -> tags
+                    .filter { it > tag1 }
```

Comment:
You are right. It is cartesian product, which generates only sorted combinations (tag1 < tag2 < tag3). Before I added this code, I checked available functions in standard library and I didn't find anything that would help. 

There are libraries that would help for sure, for example: https://arrow-kt.io/, but when you look at how to use it, I think the code is less understandable than current solution: https://stackoverflow.com/a/53750303/11056082.

Also it would be probably slower, because it is general use function and it simply generates the product. We would have to generate full cartesian product first and then filter it for combinations that we need. In current solution we generate needed combinations at once.